### PR TITLE
feat(EllipticCurve): basic API for singular cubics

### PR DESCRIPTION
**Pull request by erdOne**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/AlgebraicGeometry/EllipticCurve/VariableChange.lean [hunk 1]
```diff
@@ -189,6 +189,38 @@ lemma variableChange_a₆ : (C • W).a₆ =
     C.u⁻¹ ^ 6 * (W.a₆ + C.r * W.a₄ + C.r ^ 2 * W.a₂ + C.r ^ 3 - C.t * W.a₃ - C.t ^ 2
       - C.r * C.t * W.a₁) := rfl
 
+lemma variableChange_inv_a₁ : (C⁻¹ • W).a₁ = C.u * W.a₁ - 2 * C.s := by
+  simp [VariableChange.inv_def]
+  field_simp
+  ring
```
Comment (acmepjz) [inline Mathlib/AlgebraicGeometry/EllipticCurve/VariableChange.lean:195]: There is a terminal `simp` here (and following results). Maybe you can change them before finalizing this PR? You may use the custom simp macro (see the beginning of this file).

File: Mathlib/AlgebraicGeometry/EllipticCurve/Singular/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+
+Warning:
+Because of the rationality assumption, it is not true that every curve that is not
+nonsingular (i.e. `IsElliptic`) is `Singular`.
+(e.g. `y² = x³ - T` is not nonsingular over `𝔽₃(T)` but the cusp is in a degree `3` extension)
+
+Implementation detail:
+This contains data for defeq and computable purposes,
+but we do have an `Subsingleton` instance on it.
+Theorems who doesn't need the defeq should assume `Nonempty E.Singular` instead.
```
Comment (acmepjz) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Singular/Basic.lean:83]:
```suggestion
Theorems which doesn't need the defeq should assume `Nonempty E.Singular` instead.
```

Or you can add `abbrev IsSingular := Nonempty E.Singular`?

And maybe you should say somewhere in docstring that `Singular` includes a choice of the (usually unique) singular point.

And I don't understand the "defeq" in "This contains data for defeq and computable purposes". If this is only a `Prop` without data, then all proofs of it are defeq. So on the contrary, the introduce of data causes trouble on defeq. I understand that the data is for computable purpose, since you need a choice of singular point and tangent direction to define the isomorphism of nonsingular point to `K^\times` or `K`.

But perhaps we can refactor the design? Instead of a data version, maybe it's better to define `IsSingular E`, `IsSingularPoint E P` and  `IsSingularPointWithSlope E P α`, and all of them are `Prop`. Then your theorems and defs can use `IsSingularPoint` and `IsSingularPointWithSlope` as assumptions.

This is similar to the current design of [PowerSeries.IsWeierstrassDivisorAt](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/PowerSeries/WeierstrassPreparation.html#PowerSeries.IsWeierstrassDivisorAt). At first it was `WeierstrassDivisionData` which contains data (https://github.com/acmepjz/lean-iwasawa/commit/fe01733ee5c5959e8155e008d493090387ec65d0#diff-4d2bbef5a07d0d05b7d80fdb6c61cd0f748478f2e5ce0f182af22bdbe15b4eb9R114). But later I found that carrying data is cumbersome. So I changed it to all `Prop` design, in this way, before you actually use its defs to retrieve data, you can `rw` it freely.

Could you consider this a little bit?

File: Mathlib/AlgebraicGeometry/EllipticCurve/Singular/Basic.lean [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+  .of_variableChange f⁻¹ (by simpa)
+
+/--
+We say that a weierstrass curve is singular if it has at least one **rational** singular point.
+We show below that the singular point (if exists) is unique, see `Affine.isSingularPoint_ext`.
+
+Warning:
+Because of the rationality assumption, it is not true that every curve that is not
+nonsingular (i.e. `IsElliptic`) is `Singular`.
+(e.g. `y² = x³ - T` is not nonsingular over `𝔽₃(T)` but the cusp is in a degree `3` extension)
```
Comment (acmepjz) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Singular/Basic.lean:78]: Great observation! BTW I have a proof that over a field, a curve is not `IsElliptic` must have a rational singular point, unless for some corner cases which requires that the field is perfect. Could you add these results? They are in https://github.com/acmepjz/my-lean-test/blob/master/gtm106/weierstrass_equation/non_singular_criterion.lean.

**Review by acmepjz — COMMENTED** [review]
Thank you very much! Looks promising.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Singular/Basic.lean (hunk 3 above)
Comment (erdOne) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Singular/Basic.lean:78]: Here's the cuspidal part of the docstring that will go in the end of all this.
```
We then give the classification of singular cubics (`Δ = 0`) as follows:
1. `E.IsCuspidal`: curves with `Δ = 0` and `c₄ = 0`
   a. Not `E.Singular` (e.g. `y² = x³ + T` over `𝔽₃(T)`):
      i. For `char K = 2` this is iff either `a₄` or `a₂a₄ + a₆` is not a square.
         See `WeierstrassCurve.nonempty_singular_iff_of_isCuspidal_of_char_two`.
     ii. For `char K = 3` this is iff `b₆` is not a cube.
         See `WeierstrassCurve.nonempty_singular_iff_of_isCuspidal_of_char_three`.
    iii. For `char K ≠ 2, 3`, this cannot happen.
         See `WeierstrassCurve.singlularOfIsCuspidal`.
     iv. This also cannot happen for perfect fields.
         See `WeierstrassCurve.nonempty_singular_of_perfectField_of_isCuspidal`.
  b. `E.Singular` but not `E.SplitSingular` (e.g. `y² = x³ + Tx²` over `𝔽₂(T)`):
      i. For `char K = 2` this is iff `a₄ + a₂²` is not a fourth power.
         See `WeierstrassCurve.nonempty_splitSingular_iff_of_isCuspidal_of_char_two`.
         The singular point is given by `(√a₄, √(a₂a₄ + a₆)`).
         See `WeierstrassCurve.singularityX_sq_of_isCuspidal_of_char_two` and
             `WeierstrassCurve.singularityY_sq_of_isCuspidal_of_char_two`.
     ii. For `char K ≠ 2` this cannot happen.
         See `WeierstrassCurve.splitSinglularOfIsCuspidal`.
    iii. This also cannot happen for perfect fields.
         See `WeierstrassCurve.nonempty_splitSingular_of_perfectField_of_isCuspidal`.
  c. `E.Singular` and `E.SplitSingular`:
      The cubic is, up to change of variable, the `standardCusp K` (`y² = x³`).
      See `WeierstrassCurve.toStandardCusp_spec`.
      The singular point is given by `(-a₁²/₁₂-a₂/₃, a₁a₂/₆ + a₁³/₂₄ - a₃/₂)`.
      See `WeierstrassCurve.singlularOfIsCuspidal`.
```
I haven't finished the nodal part yet.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Singular/Basic.lean (hunk 2 above)
Comment (erdOne) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Singular/Basic.lean:83]: By "defeq" I meant that I want `(standardCusp R).singularityX = 0` to be defeq (where `standardCusp` is defined in a separate PR). Contrary to the `WeierstrassDivisionData` example you gave, the data here is essentially unique so defeq issues on the typeclass itself shouldn't happen that often (though I'm not that sure).

We do already have the `IsSingularPoint` predicate in this PR, and the `IsSingular` predicate is `Nonempty E.Singular` as you have said. So if I parsed your suggestion correctly, you are saying we shouldn't have the bundled type of rational singular point at all?

Comment (acmepjz) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Singular/Basic.lean:83]: > So if I parsed your suggestion correctly, you are saying we shouldn't have the bundled type of rational singular point at all?

You understand correctly. At least for my case the bundled `WeierstrassDivisionData` is removed since I found it cumbersome. But in your case let me think it a while. Maybe we can discuss on Zulip? Maybe reusing "thoughts on elliptic curve" thread is fine?

File: Mathlib/AlgebraicGeometry/EllipticCurve/VariableChange.lean (hunk 1 above)
Comment (erdOne) [reply Mathlib/AlgebraicGeometry/EllipticCurve/VariableChange.lean:195]: I think `simp` followed by `field_simp` and `ring` are fine according to mathlib policy. This is because `field_simp` and `ring` are versatile tactics that could absorb the variances in goals that simp-set changes might cause (compare to `rw` or `refine` where the goal needs to be in a very specific state to work).

File: Mathlib/AlgebraicGeometry/EllipticCurve/Singular/Basic.lean (hunk 3 above)
Comment (acmepjz) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Singular/Basic.lean:78]: If I remember correctly, in the nodal case, the singular point must be rational for any field.

Comment (erdOne) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Singular/Basic.lean:78]: By the way do you know what the group of (nontrivial) rational points look like in case 1a? (For example `y² = x³ + T` (where there is probably none) or `y² = x³ + T²` (where there is at least 2 and perhaps exactly 2) over `𝔽₃(T)`). I have asked @Multramate about this and he has no idea as well.

Comment (erdOne) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Singular/Basic.lean:78]: > If I remember correctly, in the nodal case, the singular point must be rational for any field.

I am fairly certain this is not true. See the examples I gave.

Comment (acmepjz) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Singular/Basic.lean:78]: Really? But I proved it here https://github.com/acmepjz/my-lean-test/blob/fd1bb144adc9ccd23794a867a2dd2cf57648b089/gtm106/weierstrass_equation/non_singular_criterion.lean#L341

Comment (acmepjz) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Singular/Basic.lean:78]: > By the way do you know what the group of (nontrivial) rational points look like in case 1a? (For example `y² = x³ + T` (where there is probably none) or `y² = x³ + T²` (where there is at least 2 and perhaps exactly 2) over `𝔽₃(T)`). I have asked @Multramate about this and he has no idea as well.

Let me think of it when I'm free.

Comment (erdOne) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Singular/Basic.lean:78]: Oh sorry I misread your "nodal" as "cuspidal". You are totally right then. I think I have some code on that as well.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Singular/Basic.lean (hunk 2 above)
Comment (erdOne) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Singular/Basic.lean:83]: Yeah maybe other people have some better opinions/suggestions. And also I can't really make good decisions now because it is 5 in the morning and I haven't slept yet 😅

File: Mathlib/AlgebraicGeometry/EllipticCurve/Singular/Basic.lean [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+
+lemma Affine.IsSingularPoint.relation_x
+    {E : WeierstrassCurve.Affine R} {x y : R} (h : E.IsSingularPoint x y) :
+    E.a₁ * y = 3 * x ^ 2 + 2 * E.a₂ * x + E.a₄ :=
+  sub_eq_zero.mp ((Affine.isSingularPoint_iff.mp h).2.1)
+
+lemma Affine.IsSingularPoint.relation_y
+    {E : WeierstrassCurve.Affine R} {x y : R} (h : E.IsSingularPoint x y) :
+    2 * y + E.a₁ * x + E.a₃ = 0 :=
+  (Affine.isSingularPoint_iff.mp h).2.2
```
Comment (Multramate) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Singular/Basic.lean:48]: Do you need these? I had `Affine.nonsingular_iff'` as the analogue of your `Affine.isSingularPoint_iff` and `Affine.nonsingular_iff` for a cleaner version without the brackets like these two lemmas.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Singular/Basic.lean [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+  cond' := by
+    rw [← singularityX]
+    letI := (inferInstanceAs (Nonempty E.Singular)).some
+    letI := (inferInstanceAs (Nonempty (f • E).Singular)).some
+    simp only [variableChange_a₁, variableChange_a₂, singularityX_variableChange,
+      VariableChange.inv_def]
+    linear_combination f.u⁻¹ ^ 2 * SplitSingular.cond E
+
+variable {E} in
+/-- The other `SplitSingular` instance obtained by swapping the two tangents -/
```
Comment (Multramate) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Singular/Basic.lean:270]:
```suggestion
/-- The other `SplitSingular` instance obtained by swapping the two tangents. -/
```

File: Mathlib/AlgebraicGeometry/EllipticCurve/Singular/Basic.lean [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+  ((E.isSingularPoint_singularity.variableChange f).ext
+      (f • E).isSingularPoint_singularity).2.symm
+
+/-- We say that a weierstrass curve is split singular if it has a (rational) singular point,
+and its tangent lines at the singular point is split in the base field.
+We later show that this is unique up to swapping the tangents.
+
+Implementation detail:
+This contains data for defeq and computable purposes. -/
+class SplitSingular [IsReduced R] (E : WeierstrassCurve R) where
```
Comment (Multramate) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Singular/Basic.lean:229]: Can this not `extend` the `Singular` class? I don't know if there is syntax for that.

**Review by Multramate — COMMENTED** [review]
Some basic comments.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Singular/Basic.lean (hunk 4 above)
Comment (erdOne) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Singular/Basic.lean:48]: This is for dot notation.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Singular/Basic.lean (hunk 6 above)
Comment (erdOne) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Singular/Basic.lean:229]: Because `Singular` is data so I chose not to.

Comment (acmepjz) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Singular/Basic.lean:229]: Maybe you can define `SplitSingular` which only contains the data `α` and the property `cond'`, but require the `E.Singular` be provided explicitly? i.e. `class SplitSingular (E : WeierstrassCurve R) [E.Singular] where ...`

Comment (erdOne) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Singular/Basic.lean:229]: I need the data for `Singular` so that the variable change will be computable. I think the most I can change here is `class SplitSingular (E : WeierstrassCurve R) [Nonempty E.Singular]` but I don't see why this would be better.

Or I just stop both of them from being a class. I haven't made up my mind yet and I planned to ask on Xena yesterday but I forgot.

Comment (acmepjz) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Singular/Basic.lean:229]: But merely `Nonempty` makes it not computable after all... you need `choice` to find out the (unique) singular point.

And may I ask what's the point to make it computable? As an example, the discriminant of an elliptic curve (as a `Units`) used to be computable, at the price that the structure `EllipticCurve` carries an extra data compared to `WeierstrassCurve` and there are a lot of API duplications. It was changed to a noncomputable version (together with predicate `IsElliptic` which is merely a `Prop`).

In your case, to define an isomorphism, one must fix a tangent direction at node; different choice of directions define different isomorphisms. So it must have some data (choose a random direction is not a good idea IMO).

In the case that the isomorphism can be defined (i.e. the base ring must be a field), the singular point must be unique, so maybe in this case we can use `Prop` instead. But then the statement of tangent direction becomes cumbersome...

Comment (acmepjz) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Singular/Basic.lean:229]: Personally I prefer unbundled `Prop` instead bundled data, i.e. instead of `[E.Singular] [E.SplitSingular]`, I prefer `IsSingularPoint E P` and `IsSingularPointWithSlope E P α`, both of them are `Prop`, and there are flexibility: user can provide the custom `P` and the proof that `P` is a singular point, or using `obtain` or `choice` to obtain it from the API `exists_isSingularPoint` we provide. For your standard singular curve, we can provide `isSingularPoint_zero`.

**Comment by leanprover-community-bot-assistant** [comment]
This pull request has conflicts, please merge `master` and resolve them.

**Comment by joneugster** [comment]
Mathlib has moved to PRs from forks. This PR is still from a branch of the main repository and will soon be closed! Please migrate the content to a fork and reopen a PR from there if you wish to do so. See [Zulip topic](https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/all.20future.20PRs.20should.20be.20made.20from.20forks/with/624081853) for more instructions. Thank you for contributing to mathlib!
